[Security] Upgrade storybook to 6.1

eatyourgreens opened the pull request:
Upgrade `@storybook` packages to 6.1.10, except for @storybook/addon-notes. `addon-notes` is deprecated at version 5.3.21.

Replace core plugins with [Storybook Essentials](https://github.com/storybookjs/storybook/tree/next/addons/essentials).
 
[Storybook 6 Migration Guide](https://storybook.js.org/blog/storybook-6-migration-guide/). 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

eatyourgreens commented:
[`@storybook/addon-docs`](https://www.npmjs.com/package/@storybook/addon-docs) is the replacement for `addon-notes`.

shaunanoordin commented:
I'm currently encountering issues with my test.

1. I've `yarn panic` and `yarn bootstrap`-ed to get a clean build
2. I've ensured I'm at commit 9516a82
3. In `package/lib-classifier`, running `yarn storybook` results in the following error in the terminal:

```
ERROR in ./.storybook/config.js-generated-config-entry.js
Module build failed (from /Users/shaun/projects/front-end-monorepo/node_modules/@storybook/core/node_modules/babel-loader/lib/index.js):
Error: /Users/shaun/projects/front-end-monorepo/packages/lib-classifier/.storybook/config.js-generated-config-entry.js: 'loose' mode configuration must be the same for @babel/plugin-proposal-class-properties, @babel/plugin-proposal-private-methods and @babel/plugin-proposal-private-property-in-object (when they are enabled).

ERROR in ./.storybook/storybook-init-framework-entry.js
Module build failed (bla bla bla same message)

ERROR in ./.storybook/addons.js
Module build failed (yadda yadda same message)

ERROR in ./.storybook/generated-refs.js
Module build failed (etc etc etc same message)
``` 

I'm going to run some tests with the other packages, then circle around back to this to see if something's up with Storybook, or if it's something local.

shaunanoordin commented:
- `package/app-project` has no issues 👍 
  - I'm able to run `yarn storybook` with no errors on the terminal
  - I'm able to open `http://localhost:9001/` on my browser and browse through some components
- `package/lib-react-components` has the same issues as lib-classifier ❌ 
  - Running `yarn storybook` results in the same errors in the terminal
  - Opening `http://localhost:6006/` in my browser results in a blank white screen, and an error message in the dev console. (Same details as terminal errors.)

eatyourgreens commented:
@shaunanoordin I've added the `loose` setting to the babel config, and migrated all the storybooks to version 6 config. They should build now.

eatyourgreens commented:
The biggest change so far seems to be that story names must use `/` as a separator.

eatyourgreens commented:
v5 notes still seem to work in storybook 6, from random testing of a few of our stories.

eatyourgreens commented:
Two patches came out since I opened this yesterday. I've bumped everything to 6.1.8.

srallen reviewed (changes requested):
A few stories are still crashing:

- Classifier storybook
  - All of the explicit dark theme stories under the banners and subject viewers crash with `Uncaught TypeError: backgrounds.find is not a function`
- React components storybook
  - Dark theme story for the GoldButton is broken too, but this is non-blocking for me since the GoldButtons should be deprecated in favor of the PrimaryButton (just hasn't been done yet)
  - Dark theme story also broken for PrimaryButton, but this should definitely be fixed
  - Non-blocking: the Tooltip component stories shows up as its own section for some reason, just a little odd.
- Project app storybook
  - Broken dark theme: http://localhost:9001/?path=/story/project-app-shared-subject-preview--dark-theme

Now these are all issues I found comparing to the current deployed storybook that seems to be caused from the update. There are some additional stories with issues, but those are present with the current deployment so need separate addressing. I don't have a strong opinion on how the dark theme stories are fixed, either as is, or replace them with the knob like how some stories are setup. Whichever takes less time is fine.

eatyourgreens commented:
6.1.10 is out now so I’ll bump this again on Monday.

eatyourgreens commented on packages/app-project/.storybook/preview.js:
```diff
@@ -0,0 +1,13 @@
+
+addParameters({
+  backgrounds: {
+    values: backgrounds.lightDefault
```
The dark theme stories need updating to use the new `backgrounds` object structure too.

eatyourgreens replied:
New backgrounds syntax: https://storybook.js.org/docs/react/essentials/backgrounds

I think this should be set in `lib/backgrounds` rather than in the preview config.

eatyourgreens commented:
I've fixed both `darkDefaults` and `lightDefaults` to use the `{ default, values }` structure for backgrounds. I've also bumped the installed packages and replaced Actions, Backgrounds and Viewports with [Storybook Essentials](https://github.com/storybookjs/storybook/tree/next/addons/essentials).

The video viewer story had the wrong title. Fixing that fixed the subject viewers.

I think `Tooltip` is set as a top-level category in the titles of the Tooltip stories.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.stories.js:
```diff
@@ -54,7 +54,7 @@ function ViewerContext (props) {
 const darkThemeConfig = Object.assign({}, config, { backgrounds: backgrounds.darkDefault })
 
-storiesOf('Subject Viewers | SingleImageViewer', module)
+storiesOf('Subject Viewers / SingleImageViewer', module)
```
It turns out the video viewer was duplicating this name. Fixing that fixed both stories.

eatyourgreens commented:
I’ve added the migration guide to the PR description too.

srallen reviewed (approved):
Stories are working now! Thank you for doing this. There's still plenty for us to do in the individual stories, but this gets us started.
